metadata/metadata_test.go

```diff
@@ -385,5 +404,197 @@ func TestTargetFilesCustomField(t *testing.T) {
 	targets.Signed.Targets["testTarget"] = targetFile
 	targetsJSON, err := targets.ToBytes(false)
 	assert.NoError(t, err)
-	assert.Equal(t, []byte("{\"signed\":{\"_type\":\"targets\",\"expires\":\"2030-08-15T14:30:45.0000001Z\",\"spec_version\":\"1.0.31\",\"targets\":{\"testTarget\":{\"custom\":{\"test\":true},\"hashes\":{},\"length\":0}},\"version\":1},\"signatures\":[]}"), targetsJSON)
+	assert.Equal(t, exampleTargetsBytes, targetsJSON)
+}
+
+func TestFromBytes(t *testing.T) {
+	safeExpiry := getSafeExpiry()
+	root := Root(safeExpiry)
+	assert.Equal(t, safeExpiry, root.Signed.Expires)
+
+	_, err := root.FromBytes(exampleRootBytes)
+	if assert.NoError(t, err) {
+		exampleRootBytesExpire := "2030-08-15T14:30:45.0000001Z"
+		expectedNewTime, err := time.Parse(time.RFC3339, exampleRootBytesExpire)
+		if assert.NoError(t, err) {
+			assert.Equal(t, expectedNewTime, root.Signed.Expires)
+		}
+	}
+}
+
+func TestToByte(t *testing.T) {
+	rootBytesExpireStr := "2030-08-15T14:30:45.0000001Z"
+	rootBytesExpire, err := time.Parse(time.RFC3339, rootBytesExpireStr)
+	if assert.NoError(t, err) {
+		root := Root(rootBytesExpire)
+		unrecognizedField := map[string]any{"test": "true"}
+		root.Signed.UnrecognizedFields = unrecognizedField
+		rootBytes, err := root.ToBytes(false)
+		assert.NoError(t, err)
+		assert.Equal(t, exampleRootBytes, rootBytes)
+	}
+}
+
+func TestFromFile(t *testing.T) {
+	safeExpiry := getSafeExpiry()
+	root := Root(safeExpiry)
+	assert.Equal(t, safeExpiry, root.Signed.Expires)
+	_, err := root.FromFile("../examples/multirepo/repository/metadata/1.root.json")
+	if assert.NoError(t, err) {
+		expectedExpire := "2024-06-06T11:12:34.649455Z"
+		expectedNewTime, err := time.Parse(time.RFC3339, expectedExpire)
+		if assert.NoError(t, err) {
+			assert.Equal(t, expectedNewTime, root.Signed.Expires)
+		}
+	}
+}
+
+func TestToFile(t *testing.T) {
+	safeExpiry := getSafeExpiry()
+	root := Root(safeExpiry)
+	tmp := os.TempDir()
+	tmpDir, err := os.MkdirTemp(tmp, "0750")
+
+	if assert.NoError(t, err) {
+		fileName := fmt.Sprintf("%s/root.json", tmpDir)
+		assert.NoFileExists(t, fileName)
+
+		err := root.ToFile(fileName, false)
+		if assert.NoError(t, err) {
+			assert.FileExists(t, fileName)
+			loadedRoot := Root(safeExpiry)
+			_, err := loadedRoot.FromFile(fileName)
+			if assert.NoError(t, err) {
+				assert.Equal(t, root.Signed.Version, loadedRoot.Signed.Version)
+				assert.Equal(t, root.Signed.Type, loadedRoot.Signed.Type)
+				assert.Equal(t, root.Signed.Expires, loadedRoot.Signed.Expires)
+			}
+		}
+		err = os.RemoveAll(tmpDir)
+		assert.NoError(t, err)
+		assert.NoFileExists(t, fileName)
+	}
+}
+
+func TestSign(t *testing.T) {
```

Comment:
There are several other things one can test for verifying the signing works correctly - are we allowed to sign that type of metadata with that given key, assert key types, if there was already a signature for that key, did we successfully ovewrote it, verify if the signature corresponds to the key and so much more.